### PR TITLE
bpo-43950: Add option to opt-out of PEP-657

### PR DESCRIPTION
**Pull request by ammaraskar**
To be merged after https://github.com/python/cpython/pull/26958

This adds the opt-out option and disables the compiler's column information and traceback printing.

https://bugs.python.org/issue43950

### CODE REVIEW

File: Lib/test/test_zipimport.py [hunk 1]
```diff
@@ -718,6 +718,7 @@ def doTraceback(self, module):
             print_tb(tb, 1, s)
             self.assertTrue(s.getvalue().endswith(
                 '    def do_raise(): raise TypeError\n'
+                '' if sys.flags.no_debug_ranges else
```
Comment (ammaraskar) [inline Lib/test/test_zipimport.py]: Not a big fan of this, there's several places in `test_traceback.py` we would have to do this. I wonder if there's a nicer way to achieve this...

Maybe some test helper that goes through and strips out lines with `*^^^^\n`?

File: Objects/codeobject.c [hunk 2]
```diff
@@ -1212,8 +1212,8 @@ code.__new__ as code_new
     name: unicode
     firstlineno: int
     linetable: object(subclass_of="&PyBytes_Type")
-    endlinetable: object(subclass_of="&PyBytes_Type")
-    columntable: object(subclass_of="&PyBytes_Type")
+    endlinetable: object
```
Comment (ammaraskar) [inline Objects/codeobject.c:1232]: Looks like we can't use `accept={NoneType}` with `object()` in argument clinic. Should we just do this manual checking or add this ability to argument clinic?

Comment (isidentical) [reply Objects/codeobject.c:1232]: I'd go with manual for now, it is not really related to this PR.

File: Lib/test/test_code.py [hunk 3]
```diff
@@ -317,6 +317,8 @@ def func():
         new_code = code = func.__code__.replace(co_linetable=b'')
         self.assertEqual(list(new_code.co_lines()), [])
 
+    @unittest.skipIf(sys.flags.no_debug_ranges,
```
Comment (isidentical) [inline Lib/test/test_code.py]: Instead of repeating this, let's make it a flag like `test.support.requires_gzip`

File: Python/sysmodule.c [hunk 4]
```diff
@@ -2580,6 +2581,7 @@ set_flags_from_config(PyInterpreterState *interp, PyObject *flags)
     SetFlagObj(PyBool_FromLong(config->dev_mode));
     SetFlag(preconfig->utf8_mode);
     SetFlag(config->warn_default_encoding);
+    SetFlagObj(PyBool_FromLong(config->no_debug_ranges));
```
Comment (pablogsal) [inline Python/sysmodule.c]: Hummm, I wonder if we should make this private for the time being

**Comment by pablogsal** [comment]
Btw, do we have a test that checks that the flag works?

**Comment by ammaraskar** [comment]
> Btw, do we have a test that checks that the flag works?

Not yet :)

File: Objects/codeobject.c (hunk 2 above)
Comment (ammaraskar) [reply Objects/codeobject.c:1232]: Sounds good.

**Comment by ammaraskar** [comment]
@pablogsal After making these changes, I am a little worried about the `sys.flags.no_debug_ranges` conditionals, I fear that without it being automatically tested somewhere we could end up with code-rot in the test suite. Would it be possible to run a buildbot with `-X no_debug_ranges` just to ensure that no bugs in the opted out version pop up.

**Comment by pablogsal** [comment]
> Would it be possible to run a buildbot with `-X no_debug_ranges` just to ensure that no bugs in the opted-out code paths pop up.

Hmmm, is possible, but I would prefer to explicitly test that the flag + env work in some specific tests using a subprocess and that would only left the `skip()` as conditional. Is true that that could break, but adding an entire buildbot is a bit of a big deal, although we can explore that.

**Comment by ammaraskar** [comment]
> Btw, do we have a test that checks that the flag works?

Just added some, please take a look.

> Hmmm, is possible, but I would prefer to explicitly test that the flag + env work in some specific tests using a subprocess

For example running `test_traceback.py` with the flag set? That would certainly work too, I like that approach.

The last remaining component for this PR is to make `marshal.c` replace the tables with `None` if the config option is set and test that.

**Comment by pablogsal** [comment]
> The last remaining component for this PR is to make `marshal.c` replace the tables with `None` if the config option is set and test that.

Maybe we should do the logic in the code object constructor and that way we don't need to deal with marshal.c and the compiler.

**Comment by ammaraskar** [comment]
> Maybe we should do the logic in the code object constructor and that way we don't need to deal with marshal.c and the compiler.

Great idea, did this in the common code path in `codeobject.c` :)

Added tests to make sure de-marshaling with the flag results in `None` for the `endlinetable` and `columntable`.

> Hmmm, is possible, but I would prefer to explicitly test that the flag + env work in some specific tests using a subprocess

Going back to this, do you know if there is an example of such a test that already exists in the suite? Where should it go? A new `test_traceback_no_debug_range.py` or similar?

**Comment by pablogsal** [comment]
> Going back to this, do you know if there is an example of such a test that already exists in the suite? Where should it go? A new `test_traceback_no_debug_range.py` or similar?

Nothing similar AFAIK, we are opening the path :). 

We should consider doing this probably in a separate PR because we would need to make sure that we don't execute everything or otherwise the test suite will start taking a lot more for not a lot of benefit. I think at the end of the day what we want to test is that the test suite doesn't break if you run it with the flag set, because many of these tests will be just skips.

**Comment by ammaraskar** [comment]
Yeah, a follow up PR after some more experimentation for that problem sounds good.

File: Lib/test/test_zipimport.py (hunk 1 above)
Comment (ammaraskar) [reply Lib/test/test_zipimport.py]: Managed to get around this by sticking a `@requires_ranges` on the test with a lot of carets and just handling the rest like this.

**Comment by ammaraskar** [comment]
@isidentical / @pablogsal Please take another look when you get the chance, this now has tests for the flag's functionality as well.

**Review by isidentical — APPROVED** [review]

File: Python/sysmodule.c (hunk 4 above)
Comment (ammaraskar) [reply Python/sysmodule.c]: Not having this here would make it a bit of a pain to test. We'd have to use something like `_testinternalcapi.get_config()`. What are the downsides to having this be public? Makes it harder to remove the opt out in the future?

Comment (pablogsal) [reply Python/sysmodule.c]: Yes, it also polutes the sys namespace. It either needs to be made private (underscore) or we need to check the config through testcapi

Comment (ammaraskar) [reply Python/sysmodule.c]: Removed from `sys.flags`, the tests now use `_testinternalcapi`.

**Comment by pablogsal** [comment]
We need more rebasing :(

**Comment by ammaraskar** [comment]
buildbot/AMD64 Ubuntu Shared PR has a weird failure...

```
/srv/buildbot/buildarea/pull_request.bolen-ubuntu/build/python: error while loading shared libraries: libpython3.11d.so.1.0: cannot open shared object file: No such file or directory
```

I wonder if we need a `@requires_*` on it or if it's just a buildbot problem?

**Comment by ammaraskar** [comment]
Nope, no `@requires` annotation is needed to use `assert_python_ok` and `assert_python_failure`. This seems to just be a problem with the buildbot.

**Comment by ammaraskar** [comment]
@pablogsal all lights are green, anything else for this one?

**Review by pablogsal — APPROVED** [review]

**Comment by pablogsal** [comment]
> @pablogsal all lights are green, anything else for this one?

Nop. Great job! 👌

**Comment by pablogsal** [comment]
@ammaraskar @isidentical Hummm, have we bumped the magic number? This buildbot failure seems quite related. I'm in some meetings right now and cannot check but may need to revert:(

**Comment by isidentical** [comment]
AFAIK we did. Checking the failures now...

**Comment by ammaraskar** [comment]
The errors on both are:

```
/srv/buildbot/buildarea/3.x.bolen-ubuntu/build/python: error while loading shared libraries: libpython3.11d.so.1.0: cannot open shared object file: No such file or directory
```

```
stderr:
---
ld-elf.so.1: Shared object "libpython3.11d.so.1.0" not found, required by "python"
---
```

when running `assert_python_ok` and `assert_python_failure` and both buildbots have "Shared" in the name.

**Comment by ammaraskar** [comment]
The confusing thing is, it only happens with the new tests, existing `assert_python_ok` tests are working just fine.

**Comment by isidentical** [comment]
To reproduce locally;
```
$ git clean -fxd
$ ./configure --with-pydebug --enable-shared -C
$ make -j
$ LD_LIBRARY_PATH=$PWD ./python -m test -v test_traceback
```

**Comment by isidentical** [comment]
We don't preserve the original environment, so shared builds fail. See here:
https://github.com/python/cpython/blob/bb3e0c240bc60fe08d332ff5955d54197f79751c/Lib/test/support/script_helper.py#L117-L121

**Comment by isidentical** [comment]
I can confirm that setting `__cleanenv` to False fixes the issue

**Comment by ammaraskar** [comment]
Aah, let's remove `__cleanenv` in that case. I set it primarily because I thought it would avoid propagation of the testing Interpreter's config.

**Comment by isidentical** [comment]
Also see this example usage of how it is used (it creates a new env based on the existing one) if you want to isolate these;

```py
    def get_hash(self, repr_, seed=None):
        env = os.environ.copy()
        env['__cleanenv'] = True  # signal to assert_python not to do a copy
                                  # of os.environ on its own
        if seed is not None:
            env['PYTHONHASHSEED'] = str(seed)
        else:
            env.pop('PYTHONHASHSEED', None)
        out = assert_python_ok(
            '-c', self.get_hash_command(repr_),
            **env)
        stdout = out[1].strip()
        return int(stdout)
```

**Pull request merged**